drivers/gpio/gpio_npm2100.c

```diff
@@ -0,0 +1,194 @@
+		if ((mask & BIT(idx)) != 0U) {
+			i2c_reg_write_byte_dt(&config->i2c, NPM2100_GPIO_OUTPUT + idx,
+					      !!(value & BIT(idx)));
+			if (ret != 0U) {
+				return ret;
+			}
+		}
+	}
+
+	return ret;
+}
+
+static int gpio_npm2100_port_set_bits_raw(const struct device *dev, gpio_port_pins_t pins)
+{
+	return gpio_npm2100_port_set_masked_raw(dev, pins, pins);
+}
+
+static int gpio_npm2100_port_clear_bits_raw(const struct device *dev, gpio_port_pins_t pins)
+{
+	return gpio_npm2100_port_set_masked_raw(dev, pins, 0U);
+}
+
+static inline int gpio_npm2100_configure(const struct device *dev, gpio_pin_t pin,
+					 gpio_flags_t flags)
+{
+	const struct gpio_npm2100_config *config = dev->config;
+	uint8_t reg = 0U;
+
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	if (pin >= NPM2100_GPIO_PINS) {
+		return -EINVAL;
+	}
+
+	/* Set initial state if defined */
+	if ((flags & (GPIO_OUTPUT_INIT_LOW | GPIO_OUTPUT_INIT_HIGH)) != 0U) {
+		int ret = i2c_reg_write_byte_dt(&config->i2c, NPM2100_GPIO_OUTPUT + pin,
+						!!(flags & GPIO_OUTPUT_INIT_HIGH));
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	/* Set pin configuration */
+	if ((flags & GPIO_INPUT) != 0U) {
+		reg |= NPM2100_GPIO_CONFIG_INPUT;
+	}
+	if ((flags & GPIO_OUTPUT) != 0U) {
+		reg |= NPM2100_GPIO_CONFIG_OUTPUT;
+	}
+	if ((flags & GPIO_SINGLE_ENDED) != 0U) {
+		reg |= NPM2100_GPIO_CONFIG_OPENDRAIN;
+	}
+	if ((flags & GPIO_PULL_UP) != 0U) {
+		reg |= NPM2100_GPIO_CONFIG_PULLUP;
+	}
+	if ((flags & GPIO_PULL_DOWN) != 0U) {
+		reg |= NPM2100_GPIO_CONFIG_PULLDOWN;
+	}
+	if ((flags & NPM2100_GPIO_DRIVE_HIGH) != 0U) {
+		reg |= NPM2100_GPIO_CONFIG_DRIVE;
+	}
+	if ((flags & NPM2100_GPIO_DEBOUNCE_ON) != 0U) {
+		reg |= NPM2100_GPIO_CONFIG_DEBOUNCE;
+	}
+
+	return i2c_reg_write_byte_dt(&config->i2c, NPM2100_GPIO_CONFIG + pin, reg);
+}
+
+static int gpio_npm2100_port_toggle_bits(const struct device *dev, gpio_port_pins_t pins)
+{
+	int ret;
+	uint32_t value;
+
+	ret = gpio_npm2100_port_get_raw(dev, &value);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return gpio_npm2100_port_set_masked_raw(dev, pins, ~value);
+}
+
+static int gpio_npm2100_pin_interrupt_configure(const struct device *dev, gpio_pin_t pin,
+						enum gpio_int_mode mode, enum gpio_int_trig trig)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(pin);
+	ARG_UNUSED(mode);
+	ARG_UNUSED(trig);
+
+	return -ENOTSUP;
+}
+
+static const struct gpio_driver_api gpio_npm2100_api = {
+	.pin_configure = gpio_npm2100_configure,
+	.port_get_raw = gpio_npm2100_port_get_raw,
+	.port_set_masked_raw = gpio_npm2100_port_set_masked_raw,
+	.port_set_bits_raw = gpio_npm2100_port_set_bits_raw,
+	.port_clear_bits_raw = gpio_npm2100_port_clear_bits_raw,
+	.port_toggle_bits = gpio_npm2100_port_toggle_bits,
+	.pin_interrupt_configure = gpio_npm2100_pin_interrupt_configure,
+};
+
+static int gpio_npm2100_init(const struct device *dev)
+{
+	const struct gpio_npm2100_config *config = dev->config;
+
+	if (!i2c_is_ready_dt(&config->i2c)) {
+		return -ENODEV;
+	}
```

Comment:
question: shouldn't register access go through MFD, so calls are _atomic_?

Reply:
I don't fully understand. Do you mean having a i2c wrapper in the mfd driver, and protect that with a mutex?